packages/components/src/disabled/index.tsx

```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import type { RefCallback } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useDisabled } from '@wordpress/compose';
+import { createContext } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { StyledWrapper } from './styles/disabled-styles';
+import type { DisabledProps } from './types';
+import type { WordPressComponentProps } from '../ui/context';
+
+const Context = createContext< boolean >( false );
+const { Consumer, Provider } = Context;
+
+/**
+ * `Disabled` is a component which disables descendant tabbable elements and prevents pointer interaction.
+ *
+ * @example
+ * ```js
+ * import { Button, Disabled, TextControl } from '@wordpress/components';
+ * import { useState } from '@wordpress/element';
+ *
+ * const MyDisabled = () => {
+ * 	const [ isDisabled, setIsDisabled ] = useState( true );
+ *
+ * 	let input = <TextControl label="Input" onChange={ () => {} } />;
+ * 	if ( isDisabled ) {
+ * 		input = <Disabled>{ input }</Disabled>;
+ * 	}
+ *
+ * 	const toggleDisabled = () => {
+ * 		setIsDisabled( ( state ) => ! state );
+ * 	};
+ *
+ * 	return (
+ * 		<div>
+ * 			{ input }
+ * 			<Button variant="primary" onClick={ toggleDisabled }>
+ * 				Toggle Disabled
+ * 			</Button>
+ * 		</div>
+ * 	);
+ * };
+ * ```
+ */
+function Disabled( {
+	className,
+	children,
+	isDisabled = true,
+	...props
+}: WordPressComponentProps< DisabledProps, 'div' > ) {
+	const ref: RefCallback< HTMLElement > = useDisabled();
+
+	if ( ! isDisabled ) {
+		return <Provider value={ false }>{ children }</Provider>;
+	}
+
+	return (
+		<Provider value={ true }>
+			<StyledWrapper
+				ref={ ref }
+				className={ classnames( className, 'components-disabled' ) }
+				{ ...props }
```

Comment:
Probably something that should not be addressed in this PR, as it's also on `trunk`, but I noticed that the remaining `...props` are only applied to the `StyledWrapper` when `isDisabled` is `true`. It would be great normalize the situation, probably the best way could be to always render a wrapper `div` getting the `...props` regardless of the value of `isDisabled` (making sure that it doesn't cause any breaking changes).

(cc in @mirka in case she has thoughts on this)

Reply:
Not a strong opinion, but I'm a bit inclined to leave it alone until we get an actual request for it. I was wondering what other prop (other than `className`) anyone would want to pass to something like this, when I found [this](https://github.com/WordPress/gutenberg/blob/6bc2f8eb51359382a7178591a430a3141a4394b7/packages/block-editor/src/components/block-preview/auto.js#L67-L69). I guess that means third-party consumers could be passing `style` props relying on the fact that it only applies when `isDisabled` is true 😅 So, until there is an actual need for this to be normalized, there might not be enough benefit to offset the breaking cost.